app/models/user.rb

```diff
@@ -28,4 +28,20 @@ def member_of?(organisation)
   def application_names
     memberships.map(&:application_names).flatten.uniq
   end
+
+  def roles_for_application(application_id)
+    memberships
+      .joins(:application_memberships)
+      .where(application_memberships: {application_id: application_id})
+      .map{|m| m.roles_for_application(application_id)}
+      .flatten.uniq
+  end
+
+  def self.current
+    Thread.current[:user]
+  end
+
+  def self.current=(user)
+    Thread.current[:user] = user
+  end
```

Comment:
This hasn't been a problem for us - presumably because we explicitly re-set User.current at the start of every request in ApplicationController ( https://github.com/ministryofjustice/ds-auth/pull/144/files#diff-55c5b7aecfb519d0e4880eaf2788eb6eR8 ) with a plain = , rather than a ||= as in your Site model.